### PR TITLE
Enable image verification upto Max Level

### LINKED ISSUE #755: Yaml Image Verify Upto max level
arkade supports container image verification for helm values.yaml. The capability should be extended to support other files such as docker-compose.yaml or any other yaml where image field is present under nested structure  
## Expected Behaviour

verify should support image validation upto max nested level, where max level can be defined by user
## Current Behaviour

The current implementation only validates upto one level below of root
## Are you a GitHub Sponsor yet (Yes/No?)

- [ ] Yes
- [x] No

## Possible Solution

1. Define a default nested level check
2. Iterate upto level and collect unique images
3. return map of images from nested level 

## Context

The current implementation is not suitable for faasd. openfaas/faasd#302

### PR DESCRIPTION
**Pull request by nitishkumar71**
Signed-off-by: Nitishkumar Singh <nitishkumarsingh71@gmail.com>

The change will allow user to validate container image present in any yaml file up to max level defined by used
## Description

1. Included a new flag where user can define max nested level to check. Default value is 3
2. Changed the current implementation to look for image key upto max level
## Motivation and Context

- [x] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))

closes #755 

## How Has This Been Tested?

1. Build in local using `make build`
2. Download [docker-compose.yaml](https://github.com/openfaas/faasd/blob/master/docker-compose.yaml) file from faasd and[ values.yaml](https://github.com/openfaas/faas-netes/blob/master/chart/cron-connector/values.yaml) file from faas-netes chart 
3. Validate images into `docker-compose.yaml` using command `./arkade chart verify -f ../faasd/docker-compose.yaml -v true`
```
Verifying images in: ../faasd/docker-compose.yaml
Found 5 images
> [ghcr.io/openfaas/queue-worker:0.12.2] ghcr.io/openfaas/queue-worker:0.12.2
< [ghcr.io/openfaas/queue-worker:0.12.2] sha256:dd69cc3d77c2e06df54ed2dddea384b6defc51ec35763a5ed377548fd30c6831
> [ghcr.io/openfaas/basic-auth:0.21.4] ghcr.io/openfaas/basic-auth:0.21.4
< [ghcr.io/openfaas/basic-auth:0.21.4] sha256:e99c385f248ce9a6569bf262e3480ddbf317cd5982243fcc229ee1766d24eb5a
> [docker.io/library/nats-streaming:0.22.0] docker.io/library/nats-streaming:0.22.0
< [docker.io/library/nats-streaming:0.22.0] sha256:448076816470a7198c7d81db17317ea061d5cb129dcbb715a046ddea3b2bf752
> [docker.io/prom/prometheus:v2.14.0] docker.io/prom/prometheus:v2.14.0
< [docker.io/prom/prometheus:v2.14.0] sha256:907e20b3b0f8b0a76a33c088fe9827e8edc180e874bd2173c27089eade63d8b8
> [ghcr.io/openfaas/gateway:0.22.0] ghcr.io/openfaas/gateway:0.22.0
< [ghcr.io/openfaas/gateway:0.22.0] sha256:57fb2e0034e879264e1dc0f8f4b6af0b803eea3a3b2e52037c0284293e473d44
```
4. Validate images into `values.yaml` using `./arkade chart verify -f ../faas-netes/chart/cron-connector/values.yaml -v true`

```
Verifying images in: ../faas-netes/chart/cron-connector/values.yaml
Found 1 images
> [ghcr.io/openfaas/cron-connector:0.6.1] ghcr.io/openfaas/cron-connector:0.6.1
< [ghcr.io/openfaas/cron-connector:0.6.1] sha256:1f33c56f972e1899a2f9abc32d7dafb14665109a82b68554adabe21467eaec9d
```

## Are you a GitHub Sponsor yet (Yes/No?)

- [ ] Yes
- [x] No

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [ ] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

File: pkg/helm/io.go [hunk 1]
```diff
@@ -28,28 +28,27 @@ func Load(yamlPath string) (ValuesMap, error) {
...
 
-		// Match anything at one level down i.e. "gateway.image: ..."
-		if c, ok := v.(map[interface{}]interface{}); ok && c != nil {
-			for kk, vv := range c {
-				if kk == "image" {
-					images[k] = vv.(string)
-				}
-			}
+		if c, ok := v.(ValuesMap); ok && maxLevel > 0 {
+			images = mergeMaps(images, FilterImagesMaxLevel(c, maxLevel-1))
```
Comment (alexellis) [inline pkg/helm/io.go:43]: Interesting use of recursion. I had to think for a moment as to what you were doing here.

File: cmd/chart/verify.go [hunk 2]
```diff
@@ -44,6 +45,7 @@ autoscaler          ghcr.io/openfaasltd/autoscaler:0.2.5
 		}
 
 		verbose, _ := cmd.Flags().GetBool("verbose")
+		maxLevel, _ := cmd.Flags().GetInt("max")
```
Comment (alexellis) [inline cmd/chart/verify.go:48]: Please validate the flag in PreRunE

Can we call "max" "depth" instead?

File: cmd/chart/verify.go [hunk 3]
```diff
@@ -36,6 +36,7 @@ autoscaler          ghcr.io/openfaasltd/autoscaler:0.2.5
 
 	command.Flags().StringP("file", "f", "", "Path to values.yaml file")
 	command.Flags().BoolP("verbose", "v", false, "Verbose output")
+	command.Flags().IntP("max", "m", 3, "Upto max level")
```
Comment (alexellis) [inline cmd/chart/verify.go:39]: `Specify how many levels deep into the YAML structure to walk looking for image: tags.`

Comment (nitishkumar71) [reply cmd/chart/verify.go]: Done

File: cmd/chart/verify.go (hunk 2 above)
Comment (nitishkumar71) [reply cmd/chart/verify.go]: Done

File: pkg/helm/io.go (hunk 1 above)
Comment (nitishkumar71) [reply pkg/helm/io.go]: Repeating code or use of something like stack to keep track of level felt a bit additional to me.

**Review by alexellis — APPROVED** [review]
Approved

**Comment by alexellis** [comment]
Would you like to add a GitHub Actions file to openfaas/faasd similar to what you can see I recently added to faas-netes to validate the images whenever the path for the docker-compose.yaml file changes?

**Comment by nitishkumar71** [comment]
Sure

**Pull request merged**